File: leetcode/024_swapnodesinpairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        curr = dummy
        
        while curr.next is not None and curr.next.next is not None:
            first = curr.next
            second = curr.next.next
            first.next = second.next
            curr.next = second
            curr.next.next = first
            curr = curr.next.next
            
        return dummy.next

File: leetcode/024_swapnodesinpairs.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
    if (!head || !head.next) return head
    let l = head
    let r = head.next

    printList(head)

    let p: ListNode | null
    head = r

    while (r) {
        // swap nodes
        if (p) p.next = r
        l.next = r.next
        r.next = l

        // printList(head)

        // move forwards
        p = l
        l = l.next
        r = l ? l.next : null
        // console.log({l: l ? l.val : "null", r: r ? r.val : "null", p: p ? p.val : "null"})
    }

    return head
};

function printList(head: ListNode | null) {
    const result = []
    while (head) {
        result.push(head.val)
        head = head.next
    }
    console.log(result)
}
